sql/expression/function/json_set.go

```diff
@@ -188,31 +186,129 @@ func (j *JSONSet) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
-func processPath(ctx *sql.Context, doc types.JSONDocument, path string, pass bool) (string, bool, error) {
-	// make sure path starts with '$'
-	if path[0] != '$' {
-		return "", false, fmt.Errorf("Invalid JSON path expression")
+func processPath(doc, path string, returnVal bool) (string, bool, bool, error) {
+	returnVal, err := checkPath(path)
+	if err != nil {
+		return "", false, false, err
 	}
+	if returnVal {
+		return "", false, true, nil
+	}
+	path = path[1:]
 
-	// no wildcards allowed in path
-	if strings.Contains(path, "*") || strings.Contains(path, "**") {
-		return "", false, fmt.Errorf("Path expressions may not contain the * and ** tokens")
+	// tokenize each field of the path
+	var parsed parsedPath
+	parsed.parts = strings.Split(path, ".")
+	if parsed.parts[0] == "" {
+		parsed.parts = parsed.parts[1:]
 	}
 
-	// if the nested path does not already exist, we do nothing
-	if len(strings.Split(path, ".")) > 2 {
-		extractedVal, _ := doc.Extract(ctx, path)
-		if extractedVal == nil {
-			pass = true
+	// process each field of the path
+	for i, part := range parsed.parts {
+		formattedPart := part
+
+		// handle any indexing in this field
+		if strings.Contains(part, "[") {
+			// tokenize indexes
+			formattedPart = strings.ReplaceAll(formattedPart, "]", "")
+			indexedParts := strings.Split(formattedPart, "[")
+			if indexedParts[0] == "" {
+				indexedParts = indexedParts[1:]
+			}
+			formattedPart = ""
+
+			// process each token
+			for j, idxPart := range indexedParts {
+				if idx, err := strconv.Atoi(idxPart); err == nil { // if token is an int, it's an index
+					if j == 0 {
+						if i == 0 {
+							return "", false, false, fmt.Errorf("ordinal indexing currently unsupported")
+						} else {
+							return "", false, false, fmt.Errorf("Invalid JSON path expression")
+						}
+					}
+
+					parentVal := gjson.Get(doc, parsed.formattedPath+formattedPart)
+
+					if !parentVal.Exists() { // if parent doesn't exist in json already, do nothing
+						return path, true, false, nil
+					}
+					if arr, ok := parentVal.Value().(map[string]interface{}); ok { // if parent is a map
+						if idx >= len(arr) {
+							return "", false, false, fmt.Errorf("index out of range for maps currently unsupported")
+						} else {
+							formattedPart = formattedPart + "." + idxPart
+						}
+					} else if arr, ok := parentVal.Value().([]interface{}); ok { // if parent is an array
+						if idx >= len(arr) { // if index out of range, append to end
+							formattedPart = formattedPart + ".-1"
+						} else {
+							formattedPart = formattedPart + "." + idxPart
+						}
+					} else {
+						if idx == 0 {
+							// if there are remaining tokens/fields, do nothing
+							if j != len(indexedParts)-1 || i != len(parsed.parts)-1 {
+								return path, true, false, nil
+							}
+						} else {
+							return "", false, false, fmt.Errorf("index out of range for single values currently unsupported")
+						}
+					}
```

Comment:
How about converting this to a switch block instead of these if/elseif/else cases? If you switch on the type of `parentVal.Value()` then I think this code will be much easier to read. It will signal to the code reader that these are logically one unit and stand out better than the if/elseif/else (especially since there are soo many if/else cases in this function). I think the code will be a little cleaner and you shouldn't need the `// if parent is map` comments, because the code will be more obvious. 

Here's an example of a golang type switch: https://go.dev/tour/methods/16